### PR TITLE
Make it easier to whitelist hangouts for restrictedadd

### PR DESCRIPTION
**Pull request by PrinterElf**
When the bot is added to a hangout with no botkeeper, or where strict_botkeeper_check is true, PM the bot admin with the details.
Echo the command to permit whitelisting the conversation so that it will remain the next time.

Add the dedicated add/remove whitelist commands.

### CODE REVIEW

**Comment by PrinterElf** [comment]
Oh, and update the help text as usual.

**Comment by zimler** [comment]
Isn't it a simple way for the botadmin to know what the convid of the HO is when getting the PM?

**Comment by PrinterElf** [comment]
@zimler There isn't, not if the bot leaves it immediately, no.

File: hangupsbot/plugins/restrictedadd.py [hunk 1]
```diff
@@ -69,12 +73,20 @@ def _check_if_admin_added_me(bot, event, command):
                     logger.warning("{} ({}) tried to add me to {}".format(
                         initiator_user_id, event.user.full_name, event.conv_id))
 
-                    yield from bot.coro_send_to_user_and_conversation(
-                        _botkeeper_list(bot, event.conv_id)[0], event.conv_id,
-                        _("<b>{}</b> ({}) tried to add me to<br/><b>{}</b><br />Run this command to enable me to stay next time:".format(event.user.full_name, initiator_user_id, bot.conversations.get_name(event.conv_id,fallback_string="? {}".format(event.conv_id)))), 
-                        _("<i>{}, you need to be authorised to add me to another conversation. I'm leaving now...</i>").format(event.user.full_name))
+                    notify = bot.config.get_option('botkeeper_check_notify')
+                    print(notify)
```
Comment (nylonee) [inline hangupsbot/plugins/restrictedadd.py]: Use logger if you want to log something, otherwise remove this line

Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: Oh, L77?

File: hangupsbot/plugins/restrictedadd.py [hunk 2]
```diff
@@ -107,9 +126,19 @@ def _verify_botkeeper_presence(bot, event, command):
     if not botkeeper:
         logger.warning("no botkeeper in {}".format(event.conv_id))
 
-        yield from bot.coro_send_message(
-            event.conv,
-            _("<i>There is no botkeeper in here. I have to go...</i>"))
+        notify = bot.config.get_option('botkeeper_check_notify')
+        if notify is True:
+            yield from bot.coro_send_to_user_and_conversation(
+                _botkeeper_list(bot, event.conv_id)[0], event.conv_id,
```
Comment (nylonee) [inline hangupsbot/plugins/restrictedadd.py]: Taking the first element from the botkepper list seems flimsy to me. Would it be better to allow admins/botkeepers the ability to opt-in to receiving the notification, rather than send it to the first botkeeper in the list?

This would involve changing the design a bit, but I think it's a better outcome.

File: hangupsbot/plugins/restrictedadd.py [hunk 3]
```diff
@@ -49,27 +53,42 @@ def _botkeeper_list(bot, conv_id):
...
             # bot was part of the event
             initiator_user_id = event.user_id.chat_id
 
-            # check if botkeeper added, including self (ie, add by link)
-            if initiator_user_id in _botkeeper_list(bot, event.conv_id):
-                logger.info("botkeeper added me to {}".format(event.conv_id))
+            #if initiator_user_id != bot_id (i.e the bot account didn't join via link)
+            if initiator_user_id != bot.user_self()["chat_id"]:
+
+                # check if botkeeper added, including self (ie, add by link)
```
Comment (nylonee) [inline hangupsbot/plugins/restrictedadd.py]: Consider reducing the number of nested if statements

File: hangupsbot/plugins/restrictedadd.py [hunk 4]
```diff
@@ -19,8 +19,12 @@ def __init__(self):
 def _initialise(bot):
     plugins.register_handler(_check_if_admin_added_me, type="membership")
     plugins.register_handler(_verify_botkeeper_presence, type="message")
-    plugins.register_admin_command(["allowbotadd", "removebotadd"])
+    plugins.register_admin_command(["allowbotadd", "removebotadd", "addhocheckwhitelist"])
 
+    if not bot.get_config_option('botkeeper_check_notify'):
```
Comment (nylonee) [inline hangupsbot/plugins/restrictedadd.py]: Consider using memory instead of config

File: hangupsbot/plugins/restrictedadd.py [hunk 5]
```diff
@@ -165,3 +187,17 @@ def removebotadd(bot, event, user_id, *args):
         yield from bot.coro_send_message(
             event.conv,
             _("user id {} is not authorised").format(user_id))
+
+
+
+def addhocheckwhitelist(bot, event, conv_id):
```
Comment (nylonee) [inline hangupsbot/plugins/restrictedadd.py]: Can we make this a toggle instead of an add-only?
It also needs a shorter and clearer name, such as `botaddalerts`

File: hangupsbot/plugins/restrictedadd.py (hunk 4 above)
Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: Used config as it's a global setting - if I use memory I then need to add a command to edit it, and we're trying to reduce those?

File: hangupsbot/plugins/restrictedadd.py (hunk 5 above)
Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: Can do - I'll just bring in the code from the earlier commit.

File: hangupsbot/plugins/restrictedadd.py (hunk 2 above)
Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: Mechanism to add?
Another command, or mandate tags?

**Comment by nylonee** [comment]
Oh. Plus wiki should be updated with the new configuration once we're closer to getting this pulled

File: hangupsbot/plugins/restrictedadd.py [hunk 6]
```diff
@@ -49,27 +53,42 @@ def _botkeeper_list(bot, conv_id):
...
-            if initiator_user_id in _botkeeper_list(bot, event.conv_id):
-                logger.info("botkeeper added me to {}".format(event.conv_id))
+            #if initiator_user_id != bot_id (i.e the bot account didn't join via link)
+            if initiator_user_id != bot.user_self()["chat_id"]:
+
+                # check if botkeeper added, including self (ie, add by link)
+                if initiator_user_id in _botkeeper_list(bot, event.conv_id):
+                    logger.info("botkeeper added me to {}".format(event.conv_id))
+
+                elif initiator_user_id is bot.user_self()["chat_id"]:
```
Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: Missed that one left over from the old code.
Removed.

File: hangupsbot/plugins/restrictedadd.py [hunk 7]
```diff
@@ -49,27 +53,42 @@ def _botkeeper_list(bot, conv_id):
...
+                    logger.info("bot added self to {}".format(event.conv_id))
+
+                else:
+                    logger.warning("{} ({}) tried to add me to {}".format(
```
Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: It is - L78 in the revised code.

File: hangupsbot/plugins/restrictedadd.py (hunk 3 above)
Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: Not sure I can - Need to check:
- Someone joined something
- Bot was involved
- Didn't add itself
- Did botkeeper add?
- If no, do you want to be notified about it?

File: hangupsbot/plugins/restrictedadd.py [hunk 8]
```diff
@@ -107,24 +127,31 @@ def _verify_botkeeper_presence(bot, event, command):
...
+                event.conv,
+                _("<i>I can't be in here without a chaperone.<br />I'm leaving now...</i>"))
 
         yield from _leave_the_chat_quietly(bot, event, command)
 
 
 @asyncio.coroutine
 def _leave_the_chat_quietly(bot, event, command):
-    yield from asyncio.sleep(1.0)
+    yield from asyncio.sleep(10.0)
```
Comment (nylonee) [inline hangupsbot/plugins/restrictedadd.py]: What's the motivation for this change?

**Comment by PrinterElf** [comment]
No idea why it was done - change was made in #740

File: hangupsbot/plugins/restrictedadd.py (hunk 8 above)
Comment (PrinterElf) [inline hangupsbot/plugins/restrictedadd.py]: No idea why it was done - change was made in #740

**Comment by endofline** [comment]
* the notification is redundant - you can redirect log messages from specific plugins with https://github.com/hangoutsbot/hangoutsbot/wiki/Logger-to-Chat-Message (re: #196) - the restrictedadd module raises a warning on unauthorised attempts which contains the conversation id:
  * https://github.com/hangoutsbot/hangoutsbot/blob/master/hangupsbot/plugins/restrictedadd.py#L65
  * https://github.com/hangoutsbot/hangoutsbot/blob/master/hangupsbot/plugins/restrictedadd.py#L108
* html shouldn't be in docstrings/comments - too verbose

**Comment by endofline** [comment]
using the tagging infrastructure (a conv tag) will also reduce the amount of code/commands required to whitelist conversations. 

in short, you "tag" a specific conversation with a keyword like "restrictedadd-whitelist", which will in turn disable botkeeper checks. the command will resemble `/bot tagset conv AAAAAAAAAAAAAAAAAAAAAAAAAA restrictedadd-whitelist`

in the plugin, check whether a conversation has the tag and act accordingly, you can retrieve tags from the bot-maintained indices: you can retrieve tags directly from the bot-maintained indices: https://github.com/hangoutsbot/hangoutsbot/wiki/Users-and-Conversations-Tagging#bottagsindices

additional references:
* the log to chat plugin mentioned in my previous comment makes use of per-conversation tags: https://github.com/hangoutsbot/hangoutsbot/blob/master/hangupsbot/commands/loggertochat.py#L110
* an example for user based tagging is available here: https://github.com/hangoutsbot/hangoutsbot/blob/master/hangupsbot/plugins/autoreply.py#L24

**Comment by endofline** [comment]
commit 59f45747449ed92f6c5ac5416be1f4e4920865cb standardises markdown syntax in docstring help, so comments can be updated for readability, the `[botalias]` replacement should still work as well

**Comment by endofline** [comment]
this PR is being closed in favor of #777 - which implements whitelisting; and also describes additional instructions how to use the logger to chat module to redirect the warning on unauthorised bot adds to any other conversation
